Implement USDT/USDC USDT LP strategy for DODO

georgeroman opened the pull request:
This PR addresses https://www.notion.so/harvestfinance/SC-DODO-finance-analyze-and-create-a-strategy-c74446ae37864fecb48591d1f4023ed4. Here is a simple analysis of DODO and the USDT/USDC V1 pool: https://www.notion.so/USDT-USDC-Dodo-Pool-4802304ee7bd4eb7b18b01a1a9c7529e.

Long story short, some pools in DODO (including the USDT/USDC V1 pool) support single asset liquidity provision. Depending on which asset you provide as liquidity, you'll get a different LP token. Any of these LP tokens can be used for liquidity mining in order to earn DODO rewards. In this particular implementation, I've chosen the USDT LP token as the underlying yield source.

DODO's rewards contract is a fork of SushiSwap's MasterChef, so most of the code is adapted from the MasterChef strategies.

Here is a test run, showing the strategy earning rewards:
```
  Mainnet USDT/USDC USDT LP
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0xFA54DAB921C65050ccb7eD5AFFFdD2178CedEb91
New Vault Deployed:  0x21E5981B31F8095766784C0691F3447fc3DAe803
Strategy Deployed:  0x580aa3Fd01D95f2058DdC67A948aE2B0E7411eD5
Strategy and vault added to Controller.
    Happy path
loop 0
old shareprice: 1000000
new shareprice: 1000000
growth: 1
loop 1
old shareprice: 1000000
new shareprice: 1000180
growth: 1.00018
loop 2
old shareprice: 1000180
new shareprice: 1000361
growth: 1.0001809674258633
loop 3
old shareprice: 1000361
new shareprice: 1000542
growth: 1.0001809346825796
loop 4
old shareprice: 1000542
new shareprice: 1000723
growth: 1.0001809019511425
loop 5
old shareprice: 1000723
new shareprice: 1000904
growth: 1.0001808692315457
loop 6
old shareprice: 1000904
new shareprice: 1001086
growth: 1.000181835620599
loop 7
old shareprice: 1001086
new shareprice: 1001267
growth: 1.000180803647239
loop 8
old shareprice: 1001267
new shareprice: 1001448
growth: 1.0001807709631896
loop 9
old shareprice: 1001448
new shareprice: 1001629
growth: 1.0001807382909547
Earned 1064595098
      ✓ Farmer should earn money (34287ms)


  1 passing (46s)
```

CryptJS13 commented:
Hey,

First of all thanks for your contribution! I have some general questions/request and I will have a look through the code you provided.

- How does the single sided LPing deal with IL? Is there a chance for users to lose a lot of their money without them expecting too?
- I see that the dodo strategy has some differences from the original MasterChef strategy, but could you use that as a base? Basically copying the `MasterChefStrategy` as `DodoSingleLPStrategy` and only making the changes necessary. That way it is easier to see the differences that are made. Also this would make it an upgradeable contract, which is prefered.
- Adding to that, keep the `DodoSingleLPStrategy` as general as possible, and put a `DodoSingleLPStrategyMainnet_USDCT_USDT` on top of it which contains the relevant parameters for that specific asset like underlying address, rewardPool address and any other addresses or variables that might be relevant for liquidation/compounding/depositing. This way it would be possible to easily reproduce the strategy for other pairs/assets and it gives a clear structure. Take a look at any of the other strategies in this repo for an example.

CryptJS13 reviewed (commented):
Few comments to incorporate in the new version

georgeroman commented:
@CryptJS13 Thanks a lot for your review! I'll get to the changes in the next couple of days.

DODO actually has one of the lowest impermanent loss in the market. As opposed to classic AMMs like Uniswap or Balancer which are based on the constant product formula, DODO uses what is called PMM (proactive market maker). Featuring an off-chain price oracle (Chainlink), DODO adjusts its pricing curve so that the price will closely follow the global market place, thus minimizing IL risk for liquidity providers. This is what allows it to support single-asset liquidity provision. Here is some more in-depth info about DODO's PMM if you want to have a deeper look: https://dodoex.github.io/docs/docs/pmmDetails#flexibility-and-k-the-liquidity-parameter.

CryptJS13 commented:
@georgeroman Alright, interesting concept, I will look into it. I don't think using oracle prices can solve IL though. Say you put 1000 USDT into a pool and someone else puts in 1000 USDC. You oracle says they are 1:1, now someone comes in and swaps 1000 USDC for 1000 USDT, now the pool has 0 USDT and 2000 USDC. You try to withdraw your USDT, but you can't because there is none.

This is a very extreme example, but in general there is a problem with single sided liquidity, as the balances in a pool will always shift, that's the whole purpose of the pool. That means liquidity providers are either locked in when they are on the wrong side of this change, or they will have to pay high fees to get out. I'll do my own research into how they handle this, but I don't see how using oracle prices solves this.

georgeroman commented:
@CryptJS13 Yes, true, this could happen, but the pool will not stay for long in that state. In such extreme scenarios, DODO would adjust its pricing (see https://dodoex.github.io/docs/docs/pmmDetails/#the-three-possible-states-in-pmm) so that it incentivizes trading back to the equilibrium (eg. global market price).

CryptJS13 commented:
@georgeroman Okay, but that does not solve IL then right? Because if the market shifts, the equilibrium position shifts with it. This won't generally be an issue for stable-stable LPs, but say you pair USDC with ETH you have IL because of the market shifts. If you put in ETH @ $2000 and then ETH price moves to $4000 dollars, the new equilibrium pool will have less ETH, so when everyone that deposited ETH wants to withdraw, this is not possible?

georgeroman commented:
@CryptJS13 When the global market price changes, DODO shifts its pricing curve accordingly so that the equilibrium price of the curve matches the global market price without having the pool balances change (this is what mitigates some of the impermanent loss). See https://dodo-in-the-zoo.medium.com/introducing-dodo-10x-better-liquidity-than-uniswap-852ce2137c57, especially the third diagram.

georgeroman commented:
@CryptJS13 I managed to get to make the changes this morning. Here's the new test result:
```bash
  Mainnet USDT/USDC USDT LP
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0xFA54DAB921C65050ccb7eD5AFFFdD2178CedEb91
New Vault Deployed:  0x21E5981B31F8095766784C0691F3447fc3DAe803
Strategy Deployed:  0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
Strategy and vault added to Controller.
    Happy path
loop 0
old shareprice: 1000000
new shareprice: 1000000
growth: 1
loop 1
old shareprice: 1000000
new shareprice: 1000179
growth: 1.000179
loop 2
old shareprice: 1000179
new shareprice: 1000358
growth: 1.0001789679647344
loop 3
old shareprice: 1000358
new shareprice: 1000537
growth: 1.0001789359409332
loop 4
old shareprice: 1000537
new shareprice: 1000716
growth: 1.0001789039285904
loop 5
old shareprice: 1000716
new shareprice: 1000895
growth: 1.0001788719276998
loop 6
old shareprice: 1000895
new shareprice: 1001073
growth: 1.0001778408324549
loop 7
old shareprice: 1001073
new shareprice: 1001252
growth: 1.000178808138867
loop 8
old shareprice: 1001252
new shareprice: 1001430
growth: 1.0001777774226668
loop 9
old shareprice: 1001430
new shareprice: 1001607
growth: 1.0001767472514305
Earned 1049878681
      ✓ Farmer should earn money (32737ms)


  1 passing (41s)
```

georgeroman commented on contracts/strategies/dodo/DodoStrategyMainnet_USDT_USDC_USDT.sol:
```diff
@@ -0,0 +1,278 @@
+        // Add this check here as well for gas efficiency
+        require(!pausedInvesting, "Investing is paused");
+
+        exitRewardPool();
```
Looks like the DODOMine contract has a `claim` method and I used that.

georgeroman commented on contracts/strategies/dodo/DodoStrategyMainnet_USDT_USDC_USDT.sol:
```diff
@@ -0,0 +1,278 @@
+        );
+
+        // Don't mark the DODO token as unsalvagable in order to be able to liquidate externally
+        unsalvagableTokens[dodo] = false;
```
What would happen in an emergency exit situation? If the reward token is unsalvageable then it's stuck in the strategy until investing is unpaused - marking it as salvageable would let governance get the rewards in this case. Is my reasoning correct here?

CryptJS13 commented on contracts/base/dodo-base/DodoV1SingleLPStrategy.sol:
```diff
@@ -0,0 +1,402 @@
+            return;
+        }
+
+        // TODO: Should we liquidate DODO to USDT on DODO first
```
This might be a good idea. Best is to do this swap before the `notifyProfitInRewardToken()`. Then USDT will be the `rewardToken` for the strategy and you notify the USDT balance. It will be better for the buyback returns.

CryptJS13 reviewed (commented):
Looks good, changing the rewardToken to USDT might be a good idea because DODO liquidity on Uni is so small.

georgeroman commented:
@CryptJS13 Addressed the comment - changed the reward token to USDT and had the DODO rewards swapped to USDT via DODO's highly liquid DODO/USDT pool when liquidating. Here are the new test results:

```bash
  Mainnet USDT/USDC USDT LP
Impersonating...
0xf00dD244228F51547f0563e60bCa65a30FBF5f7f
0xFA54DAB921C65050ccb7eD5AFFFdD2178CedEb91
New Vault Deployed:  0x21E5981B31F8095766784C0691F3447fc3DAe803
before [ '0xc95CbE4ca30055c787CB784BE99D6a8494d0d197',
  '0x21E5981B31F8095766784C0691F3447fc3DAe803' ] {}
Strategy Deployed:  0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
Strategy and vault added to Controller.
    Happy path
loop 0
old shareprice: 1000000
new shareprice: 1000000
growth: 1
loop 1
old shareprice: 1000000
new shareprice: 1000180
growth: 1.00018
loop 2
old shareprice: 1000180
new shareprice: 1000361
growth: 1.0001809674258633
loop 3
old shareprice: 1000361
new shareprice: 1000542
growth: 1.0001809346825796
loop 4
old shareprice: 1000542
new shareprice: 1000723
growth: 1.0001809019511425
loop 5
old shareprice: 1000723
new shareprice: 1000904
growth: 1.0001808692315457
loop 6
old shareprice: 1000904
new shareprice: 1001085
growth: 1.0001808365237825
loop 7
old shareprice: 1001085
new shareprice: 1001267
growth: 1.0001818027440228
loop 8
old shareprice: 1001267
new shareprice: 1001448
growth: 1.0001807709631896
loop 9
old shareprice: 1001448
new shareprice: 1001629
growth: 1.0001807382909547
Earned 1064472831
      ✓ Farmer should earn money (170059ms)


  1 passing (3m)
```

CryptJS13 reviewed (commented):
Looks good!

breadforthepeople commented:
Communicated via DMs on discord. Closing for the time being.
